server_development/topics/auth-spi.adoc

```diff
@@ -108,199 +111,547 @@ Let's walk through the steps from when a client first redirects to keycloak to a
   This requirement is satisfied because the UsernamePassword provider already associated the user with the flow.
+  This provider's `matchCondition` method will evaluate the `configuredFor` method for all other Authenticators in its current subflow. If the subflow contains
+  executors with their Requirement set to required, then the `matchCondition` method will only evaluate to true if all the required authenticators' `configuredFor`
+  method evaluate to true. Otherwise, the `matchCondition` method will evaluate to true if any alternative authenticator evaluates to true.
+. The next execution is the OTP Form.
+  This also provider requires that a user has been associated with the flow.
+  This requirement is satisfied because the UsernamePassword provider already associated the user with the flow.
   Since a user is required for this provider, the provider is also asked if the user is configured to use this provider.
-  If user is not configured, and this execution is required, then the flow will then set up a required action that the user must perform after authentication is complete.
-  For OTP, this means the OTP setup page.
-  If the execution was optional, then this execution is skipped. 
+  If user is not configured, then the flow will then set up a required action that the user must perform after authentication is complete.
+  For OTP, this means the OTP setup page. If the user is configured, he will be asked to enter his otp code. In our scenario, because of the conditional
+  sub-flow, the user will never see the OTP login page, unless the Conditional OTP subflow is set to Required.
 . After the flow is complete, the authentication processor creates a UserSessionModel and associates it with the AuthenticationSessionModel.
-  It then checks to see if the user is required to complete any required actions before logging in. 
+  It then checks to see if the user is required to complete any required actions before logging in.
 . First, each required action's evaluateTriggers() method is called.
   This allows the required action provider to figure out if there is some state that might trigger the action to be fired.
-  For example, if your realm has a password expiration policy, it might be triggered by this method. 
+  For example, if your realm has a password expiration policy, it might be triggered by this method.
 . Each required action associated with the user that has its requiredActionChallenge() method called.
   Here the provider sets up an HTTP response which renders the page for the required action.
-  This is done by setting a challenge status. 
-. If the required action is ultimately successful, then the required action is removed from the user's require actions list. 
-. After all required actions have been resolved, the user is finally logged in.         
+  This is done by setting a challenge status.
+. If the required action is ultimately successful, then the required action is removed from the user's require actions list.
```

Comment:
should "user's require" be "user's required" ?

Reply:
Yes, will correct.